File: src/main/java/cs/ut/controllers/modal/ParameterModalController.java
package cs.ut.controllers.modal;

import com.google.common.collect.Lists;
import com.google.common.escape.Escaper;
import com.google.common.html.HtmlEscapers;
import cs.ut.config.MasterConfiguration;
import cs.ut.controllers.MainPageController;
import cs.ut.engine.JobManager;
import cs.ut.engine.NirdizatiThreadPool;
import cs.ut.jobs.DataSetGenerationJob;
import cs.ut.ui.GridValueProvider;
import cs.ut.ui.NirdizatiGrid;
import cs.ut.ui.providers.ColumnRowValueProvider;
import cs.ut.util.ConstKt;
import cs.ut.util.CsvReader;
import org.apache.log4j.Logger;
import org.zkoss.bind.annotation.AfterCompose;
import org.zkoss.util.resource.Labels;
import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.event.Event;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.event.SerializableEventListener;
import org.zkoss.zk.ui.select.annotation.Wire;
import org.zkoss.zk.ui.util.Clients;
import org.zkoss.zk.ui.util.GenericAutowireComposer;
import org.zkoss.zul.Button;
import org.zkoss.zul.Combobox;
import org.zkoss.zul.Comboitem;
import org.zkoss.zul.Hlayout;
import org.zkoss.zul.Label;
import org.zkoss.zul.Row;
import org.zkoss.zul.Rows;
import org.zkoss.zul.Window;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ParameterModalController extends GenericAutowireComposer<Component> {
    private static final Logger log = Logger.getLogger(ParameterModalController.class);

    @Wire
    private Window modal;

    @Wire
    private Hlayout gridSlot;

    private NirdizatiGrid<String> grid;

    @Wire
    private Button cancelBtn;

    @Wire
    private Button okBtn;

    private Rows rows;

    private List<String> cols = MasterConfiguration.getInstance().getCSVConfiguration().getUserCols();

    private List<Combobox> fields = new ArrayList<>();

    private SerializableEventListener<Event> okBtnListener;

    private Map<String, List<String>> identifiedColumns;

    private File file;

    private transient CsvReader csvReader;

    private static final String ignoreColumn = "params.ignore";

    @AfterCompose
    public void doAfterCompose(Component comp) throws Exception {
        super.doAfterCompose(comp);

        this.file = (File) arg.get("file");
        csvReader = new CsvReader(file);

        log.debug(String.format("Modal received file <%s>", file));
        log.debug(String.format("Current log file : <%s>", file.getName()));
        List<String> fileColumns = csvReader.readTableHeader();
        log.debug(String.format("Columns present in table: <%s>", fileColumns));
        Collections.sort(fileColumns);

        Map<String, String> identifiedCols = new HashMap<>();
        csvReader.identifyUserColumns(fileColumns, identifiedCols);

        Escaper escaper = HtmlEscapers.htmlEscaper();

        GridValueProvider<String, Row> provider = new ColumnRowValueProvider(fileColumns, identifiedCols);
        grid = new NirdizatiGrid<>(provider);
        grid.setHflex("min");

        if (fileColumns.isEmpty()) {
            Clients.showNotification(Labels.getLabel(
                    "modals.unknown_separator",
                    new Object[]{escaper.escape(file.getName())}),
                    "error", getPage().getFirstRoot(),
                    "bottom_center",
                    -1);

            JobManager.Manager.flushJobs();
            modal.detach();
            return;
        }

        grid.generate(cols, true);

        cancelBtn.addEventListener(Events.ON_CLICK, (SerializableEventListener<Event>) e -> {
            Files.delete(Paths.get(file.getAbsolutePath()));
            modal.detach();
        });

        okBtnListener = e -> {
            okBtn.setDisabled(true);
            updateContent(csvReader.generateDatasetParams(grid.gatherValues()));
        };

        okBtn.addEventListener(Events.ON_CLICK, okBtnListener);

        gridSlot.appendChild(grid);
        log.debug("Showing modal");
    }

    private void updateContent(Map<String, List<String>> parameters) {
        okBtn.setDisabled(false);
        identifiedColumns = parameters;
        fields = new ArrayList<>();

        modal.setTitle(Labels.getLabel("modals.confirm_columns"));

        okBtn.removeEventListener(Events.ON_CLICK, okBtnListener);
        okBtnListener = e -> {
            Map<String, List<String>> acceptedParameters = gatherAcceptedValues();
            acceptedParameters.forEach((k, v) -> identifiedColumns.put(k, v));
            new NirdizatiThreadPool().execute(new DataSetGenerationJob(identifiedColumns, file, execution.getDesktop()));
            Clients.showNotification(Labels.getLabel("upload.success", new Object[]{HtmlEscapers.htmlEscaper().escape(file.getName())}), "info", getPage().getFirstRoot(), "bottom_right", -1);
            MainPageController.getInstance().setContent(ConstKt.PAGE_TRAINING, getPage());
            modal.detach();
        };

        okBtn.addEventListener(Events.ON_CLICK, okBtnListener);

        rows = new Rows();
        grid.getChildren().clear();
        grid.appendChild(rows);
        grid.setMold("paging");
        grid.setPageSize(10);

        Escaper escaper = HtmlEscapers.htmlEscaper();
        List<String> changeableVals = csvReader.getColumnList();
        changeableVals.sort(String::compareToIgnoreCase);

        changeableVals.forEach(key -> {
            List<String> columns = identifiedColumns.get(key);
            columns.sort(String::compareToIgnoreCase);

            columns.forEach(col -> {
                Row row = new Row();
                row.appendChild(new Label(escaper.escape(col)));

                Combobox combobox = new Combobox();
                combobox.setReadonly(true);
                combobox.setId(col);
                combobox.setConstraint("no empty");

                changeableVals.forEach(item -> {
                    Comboitem comboitem = combobox.appendItem(Labels.getLabel("params.".concat(item)));
                    comboitem.setValue(item);
                    if (item.equalsIgnoreCase(key)) combobox.setSelectedItem(comboitem);
                });

                Comboitem comboitem = combobox.appendItem(Labels.getLabel(ignoreColumn));
                comboitem.setValue(ignoreColumn);

                row.appendChild(combobox);
                rows.appendChild(row);
                fields.add(combobox);
            });
        });
    }

    private Map<String, List<String>> gatherAcceptedValues() {
        Map<String, List<String>> vals = new HashMap<>();

        fields.forEach(field -> {
            if (!ignoreColumn.equalsIgnoreCase(field.getSelectedItem().getValue())) {
                if (!vals.containsKey(field.getSelectedItem().getValue())) {
                    vals.put(field.getSelectedItem().getValue(), Lists.newArrayList(field.getId()));
                } else {
                    vals.get(field.getSelectedItem().getValue()).add(field.getId());
                }
            }
        });

        csvReader.getColumnList().forEach(col -> {
            if (!vals.containsKey(col)) {
                vals.put(col, new ArrayList<>());
            }
        });

        return vals;
    }
}

File: src/main/kotlin/cs/ut/controllers/modal/ParameterModalController.kt
package cs.ut.controllers.modal

import com.google.common.html.HtmlEscapers
import cs.ut.config.MasterConfiguration
import cs.ut.controllers.MainPageController
import cs.ut.engine.JobManager
import cs.ut.engine.NirdizatiThreadPool
import cs.ut.jobs.DataSetGenerationJob
import cs.ut.ui.NirdizatiGrid
import cs.ut.ui.providers.ColumnRowValueProvider
import cs.ut.ui.providers.ComboArgument
import cs.ut.ui.providers.ComboProvider
import cs.ut.util.CsvReader
import cs.ut.util.TIMESTAMP_COL
import org.apache.log4j.Logger
import org.zkoss.util.resource.Labels
import org.zkoss.zk.ui.Component
import org.zkoss.zk.ui.event.Event
import org.zkoss.zk.ui.event.Events
import org.zkoss.zk.ui.event.SerializableEventListener
import org.zkoss.zk.ui.select.annotation.Wire
import org.zkoss.zk.ui.util.Clients
import org.zkoss.zk.ui.util.GenericAutowireComposer
import org.zkoss.zul.Button
import org.zkoss.zul.Hlayout
import org.zkoss.zul.Window
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths

class ParameterModalController : GenericAutowireComposer<Component>() {
    private val log: Logger = Logger.getLogger(ParameterModalController::class.java)!!

    @Wire
    private lateinit var modal: Window

    @Wire
    private lateinit var gridSlot: Hlayout

    @Wire
    private lateinit var cancelBtn: Button

    @Wire
    private lateinit var okBtn: Button

    private lateinit var cols: List<String>

    private lateinit var okBtnListener: SerializableEventListener<Event>

    private lateinit var file: File

    private lateinit var csvReader: CsvReader

    companion object {
        const val IGNORE_COL = "ignore"
    }

    override fun doAfterCompose(comp: Component?) {
        super.doAfterCompose(comp)

        cols = MasterConfiguration.getInstance().csvConfiguration.userCols
        log.debug("Read columns from master config: $cols")

        this.file = arg["file"] as File
        csvReader = CsvReader(file)

        log.debug("Received file with name ${file.name}")

        val header = csvReader.readTableHeader().sorted()
        log.debug("Read header of users file: $header")

        if (validateDataPresent(header)) return

        val identifiedColumns = mutableMapOf<String, String>()
        csvReader.identifyUserColumns(header.toMutableList(), identifiedColumns)
        identifiedColumns[TIMESTAMP_COL] = csvReader.getTimeStamp()

        val provider = ColumnRowValueProvider(header, identifiedColumns)
        val grid = NirdizatiGrid(provider)
        grid.hflex = "min"

        grid.generate(cols)

        cancelBtn.addEventListener(Events.ON_CLICK, { _ ->
            Files.delete(Paths.get(file.absolutePath))
            modal.detach()
        })

        okBtnListener = SerializableEventListener { _ ->
            okBtn.isDisabled = true
            updateContent(csvReader.generateDatasetParams(grid.gatherValues()))
        }

        okBtn.addEventListener(Events.ON_CLICK, okBtnListener)

        gridSlot.appendChild(grid)
        log.debug("Log parsing successful, showing modal")
    }

    @SuppressWarnings("unchecked")
    private fun updateContent(params: MutableMap<String, MutableList<String>>) {
        okBtn.isDisabled = false
        log.debug("Updating content with params $params")

        val grid = prepareGrid()

        modal.title = Labels.getLabel("modals.confirm_columns")
        okBtn.removeEventListener(Events.ON_CLICK, okBtnListener)
        log.debug("Removed ok button listener")

        okBtnListener = SerializableEventListener { _ ->
            val accepted = grid.gatherValues() as Map<String, String>
            accepted.forEach { k, v ->
                params.values.forEach {
                    if (k in it) it.remove(k)
                }

                if (v in params) {
                    params[v]!!.add(k)
                } else {
                    params[v] = mutableListOf(k)
                }
            }
            NirdizatiThreadPool().execute(DataSetGenerationJob(params, file, execution.desktop))
            Clients.showNotification(Labels.getLabel("upload.success", arrayOf(HtmlEscapers.htmlEscaper().escape(file.getName()))), "info", getPage().getFirstRoot(), "bottom_center", -1)
            MainPageController.getInstance().setContent("landing", getPage())
            modal.detach()
        }
        okBtn.addEventListener(Events.ON_CLICK, okBtnListener)

        val escaper = HtmlEscapers.htmlEscaper()
        var args = listOf<ComboArgument>()
        val changeable: List<String> = csvReader.getColumnList().sortedBy { it.toLowerCase() }
        changeable.forEach { key ->
            val cols = params[key]!!

            cols.forEach {
                args += ComboArgument(escaper.escape(it), changeable + IGNORE_COL, key)
            }
        }

        grid.generate(args)
    }

    private fun prepareGrid(): NirdizatiGrid<ComboArgument> {
        val grid = NirdizatiGrid(ComboProvider())
        gridSlot.getChildren<Component>().clear()
        gridSlot.getChildren<Component>().add(grid)

        grid.setMold("paging")
        grid.setPageSize(10)
        grid.hflex = "min"

        return grid
    }


    private fun validateDataPresent(header: List<String>): Boolean {
        if (header.isEmpty()) {
            Clients.showNotification(
                    Labels.getLabel("modals.unknown_separator",
                            arrayOf(HtmlEscapers.htmlEscaper().escape(file.name))),
                    "error", getPage().firstRoot,
                    "bottom_center"
                    , -1, true
            )
            JobManager.flushJobs()
            modal.detach()
            return true
        }
        return false
    }
}